packages/chakra-components/src/MultiSelect/useComboBox.js

```diff
@@ -0,0 +1,346 @@
+import { useState, useEffect, useRef, useCallback } from "react";
+import debounce from "lodash.debounce";
+import { useForkRef } from "@chakra-ui/core/dist/utils";
+import { useConstant } from "./useConstant";
+
+const defaultFilter = (options, input) => {
+  if (input) {
+    return options.filter(option =>
+      option.label.toLowerCase().includes(input.toLowerCase()),
+    );
+  } else {
+    return options;
+  }
+};
+
+export const useComboBox = ({
+  options,
+  value: initialValues,
+  onChange,
+  filteredBy = defaultFilter,
+  isMulti,
+  isListBox,
+}) => {
+  // Refs
+  const listRef = useRef(null);
+  const inputRef = useRef(null);
+  const multiSelectRef = useRef(null);
+  const popperRef = useRef(null);
+
+  let _initialValues = [];
+
+  // TODO: Check if this calculation for the initial value is battle proven
+  if (initialValues) {
+    if (Array.isArray(initialValues)) {
+      _initialValues = initialValues;
+    } else {
+      if (initialValues) _initialValues = [initialValues];
+    }
+  }
+
+  // States
+  const [values, setValues] = useState(_initialValues);
+  const [selectedOptions, setSelectedOptions] = useState([]);
+  const [notSelectedOptions, setNotSelectedOptions] = useState(options);
+  const [filteredOptions, setFilteredOptions] = useState(notSelectedOptions);
+  const [inputValue, setInputValue] = useState("");
+  const [listBoxInputValue, setListBoxInputValue] = useState("");
+  const [inputIsHidden, setInputIsHidden] = useState(false);
+  const [isFocused, setIsFocused] = useState(false);
+  const [isOpen, setIsOpen] = useState(false);
+  const [focusedOptionIndex, setFocusedOptionIndex] = useState(0);
```

Comment:
Would prefer to use useReducer. Closure issue will happen when we do multiple state updates in a single cb.

Reply:
Got it.